### PR TITLE
Redux store

### PR DESCRIPTION
**Pull request by ShelOrock**
Redux store stuff. Added constants, actions, thunks, reducers. Combined all reducers. Wrapped app in a Provider with redux store. Updated api routes to send created/updated users to use on the frontend. Updated api routes to have our agreed upon names (cart -> orderDetails)

TODO: connect a component to the store to test.
TODO: Wishlist constants, actions, thunks, reducers.

Questions:
1. Some thunks dispatch another thunk (like creating a new product and then fetching the updated list of products). Do you think two api calls in some actions is $$$ or is this fine for now?
2. Thunks are repetitive. Do you think there are ways to DRY them out?
3. How should we handle errors here? I'm only console logging, but some could feasibly reset that part of the store or something.

### CODE REVIEW

File: app/redux/thunks.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+import {
+    setProducts,
+    setSingleProduct,
+    setUser,
+    setOrders,
+} from 'actions.js';
+
+//Thunk for fetching all products. 
+//Do you think we could initialize id to null, and reuse this thunk to make a call for a single product?
```
Comment (nries1) [inline app/redux/thunks.js]: We probably could. But I don't have a problem with using two thunks for this.

File: app/redux/thunks.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        .catch(e => console.error('Error fetching single product', e))
+    }
+}
+
+//Thunk for creating a product.
+//Refetches products after creating.
+export const postProduct = product => {
+    return dispatch => {
+        return axios.post(`/api/products`, product)
+        .then(() => dispatch(fetchProducts()))
```
Comment (nries1) [inline app/redux/thunks.js]: My redux isn't great, but is this meant to dispatch a function that dispatches? Can you just use .then(() => fetchProducts()) ?

File: app/redux/thunks.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        .catch(e => console.error('Error updating product', e))
+    }
+}
+
+//Thunk for fetching a user
+export const fetchUser = (userId = null) => {
+    return dispatch => {
+        return axios.get(`/api/users/${ userId }`)
+        .then(res => dispatch(setUser(res.data)))
+        .catch(e => console.error('Error fetching user', e))
```
Comment (nries1) [inline app/redux/thunks.js]: Good point about error handling. If the error occurs during a read or write action to the server, then we may want to create an "error component" that displays when there is a critical error.

File: app/redux/thunks.js (hunk 2 above)
Comment (ShelOrock) [reply app/redux/thunks.js]: Make sense. Will try it.

File: app/redux/reducers.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        case SET_ORDERS:
+            return action.orders;
+        default:
+            return state;
+    }
+}
+
+export const orderDetails = (state = [], action) => {
+    switch(action.type) {
+        case SET_ORDER_DETAILS:
```
Comment (raymondng1) [inline app/redux/reducers.js:47]: Looks like we need to import ' SET_ORDER_DETAILS ' from './constants' as well.

File: app/redux/thunks.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        .then(res => dispatch(setOrders(res.data)))
+        .catch(e => console.error('Error fetching orders', e))
+    }
+}
+
+//Thunk for creating a new order.
+//Refetches orders after creating.
+export const postOrder = (userId, order) => {
+    return dispatch => {
+        return axios.post(`/api/${ userId }/orders`, order)
```
Comment (nries1) [inline app/redux/thunks.js]: How does this route get activated?

which middleware is listening for `/api/:id/orders`? If it's in the users routes, then shouldn't this be `/api/users/${userId}/orders` ?

File: app/store.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { createStore, applyMiddleware } from 'redux';
+import { createLogger } from 'redux-logger';
+import thunkMiddleware from 'redux-thunk';
+import axios from 'axios';
+
```
Comment (raymondng1) [inline app/store.js:6]: maybe we can import some redux dev tools for easier redux debugging such as: 

createLogger from 'redux-logger';
composeWithDevTools from 'redux-devtools-extension';

File: app/redux/reducers.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            return action.products;
+        default:
+            return state;
+    }
+}
+
+export const singleProduct = (state = {}, action) => {
+    switch(action.type) {
+        case SET_SINGLE_PRODUCT:
+            return action.user;
```
Comment (raymondng1) [inline app/redux/reducers.js]: return `action.product` ?

**Review by raymondng1 — COMMENTED** [review]
Great job on the redux! The changes and everything looks great to me

**Review by raymondng1 — DISMISSED** [review]

**Review by nries1 — APPROVED** [review]

**Pull request merged**